### PR TITLE
CPU flash attention with mask

### LINKED ISSUE #112380: Need support CPU flash attention with mask
### 🚀 The feature, motivation and pitch

I want use flash attention on the cpu to speed up my automatic speech recognition(ASR) model, which is transformer-based.
In practice, the audio length in a batch may be different, which corresponds to different sequence lengths. The general approach is to use a mask tensor to indicate the sequence length, and add the mask to the attenton score before calculating the softmax in SDPA. Here is a simple example:
```python
def get_mask(self, lengths, dtype=torch.float32):
    # lengths: the seq's length, shape is [batch, ]
    max_seq_len = lengths.max().item()
    row_vector = torch.arange(0, max_seq_len, 1)
    matrix = torch.unsqueeze(lengths, dim=-1)
    mask = row_vector < matrix
    mask = mask.to(dtype)
    mask_4d = 1 - mask[:, None, None, :]
    mask_4d = mask_4d * torch.finfo(dtype).min
    return mask_4d, mask

lengths = [2, 3]  # batch size is 2, first seq length is 2, second is 3
mask_4d, mask = get_mask(lengths)
print(mask)
"""
[
    [1, 1, 0],
    [1, 1, 1],
]
"""

# Assume here is a sdpa block
# query, key, value are both [batch, num_heads, max_seq_len, head_dim]
q_h = query * head_dim ** (-0.5)
attention_scores = torch.matmul(q_h, key.transpose(-1, -2))
attention_scores = attention_scores + mask_4d
attention_probs = nn.functional.softmax(attention_scores, dim=-1)
context_layer = torch.matmul(attention_probs, value)
```
As the code shows, mask4d needs to be used before softmax to eliminate erroneous results.
For now, I can use `torch._scaled_dot_product_attention_math` for this, but `torch._scaled_dot_product_flash_attention` doesn't work, it doesn't provide an api for mask.

However, when I see `flash_attention_kernel` in `FlashAttentionKernel.cpp`, which is the kernel for cpu flash attention, it could support mask just simply modify the code.

### Alternatives

_No response_

### Additional context

_No response_

cc @jbschlosser @bhosmer @cpuhrsch @erichan1 @drisspg @mikaylagawarecki

### PR DESCRIPTION
**Pull request by imzhuhl**
Fixes #112380

According to the issue, I try to add mask support to the flash attn structure on CPU.

First mask is supported by using cum_seq_[q/kv], so the code for flashAttentionKernel is modified.
Secondly, an interface needs to be added to allow users to call, I choose to overload the `_scaled_dot_product_flash_attention` function.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Comment by drisspg** [comment]
I think we want to do this, we need to properly spit the APIs I.e. add this: https://github.com/pytorch/pytorch/pull/110546/

And this and then you would add an impl to the "nested" registration for CPU

cc @cpuhrsch

**Comment by imzhuhl** [comment]
> I think we want to do this, we need to properly spit the APIs I.e. add this: #110546
> 
> And this and then you would add an impl to the "nested" registration for CPU
> 
> cc @cpuhrsch

So we expect pytorch to have two different flash attn APIs. I think I should commit the changes after https://github.com/pytorch/pytorch/pull/110546 merges, Is that okay?

**Comment by Valentine233** [comment]
For the support of attention mask, I suppose that using `cum_seq_q` or `cum_seq_k` is not appropriate because they do not mean the same thing as attention mask.

Maybe it's better to implement `_scaled_dot_product_efficient_attention` CPU version as it has `attn_bias` as input? I could help do this if needed.

cc @jgong5

**Comment by jgong5** [comment]
> For the support of attention mask, I suppose that using `cum_seq_q` or `cum_seq_k` is not appropriate because they do not mean the same thing as attention mask.
> 
> Maybe it's better to implement `_scaled_dot_product_efficient_attention` CPU version as it has `attn_bias` as input? There are some early discussions here #103826. I could help do this if needed.
> 
> cc @jgong5

Agreed with @Valentine233 . Overloading the meaning of `cum_seq_q` or `cum_seq_k` for attention mask doesn't seem the right way to support the mask. Since the algorithms for flash attention and efficient attention are very similar, I don't see the benefit of implement another version for efficient attention on CPU. Would it be a viable option to provide an overload for `_scaled_dot_product_flash_attention` to support attention mask with an added arg for `attn_mask` on CPU? @drisspg

**Comment by Valentine233** [comment]
Hi @drisspg, do you have any opinion on the API to support CPU flash attention with mask?

**Comment by drisspg** [comment]
Hey, so like @jgong5 said I think it doesn't really make sense too add a new backend if we are going to be modifying the existing kernel code. 

If it makes it easier we can decouple the signature of the kernel from cpu and cuda. We should probably register a new stem function could and call it "sdpa_fused_attention" or something where we only register a cpu backend. 

I think this would also clean up the meta registration for sdpa as we won't be forcing the cpu code to abide by the constraints of the cuda code
